### PR TITLE
Service for custom social network icons

### PR DESCRIPTION
**Pull request by vorfolomey**

### CODE REVIEW

File: core/src/main/java/greencity/dto/user/UserProfileDtoResponse.java [hunk 1]
```diff
@@ -14,7 +16,7 @@
     private String firstName;
     private String city;
     private String userCredo;
-    private List<String> socialNetworks;
+    private List<SocialNetworkResponseDTO> socialNetworks = new ArrayList<>();
```
Comment (antonbakalets) [inline core/src/main/java/greencity/dto/user/UserProfileDtoResponse.java:19]: Field initialization is the Mapper responsibility.

File: core/src/main/java/greencity/mapping/MultipartBase64ImageMapper.java [hunk 2]
```diff
@@ -40,7 +37,14 @@ protected MultipartFile convert(String image) {
             ImageIO.write(bufferedImage, "png", tempFile);
             FileItem fileItem = new DiskFileItem("mainFile", Files.probeContentType(tempFile.toPath()),
                 false, tempFile.getName(), (int) tempFile.length(), tempFile.getParentFile());
-            fileItem.getOutputStream();
+            InputStream input = new FileInputStream(tempFile);
```
Comment (antonbakalets) [inline core/src/main/java/greencity/mapping/MultipartBase64ImageMapper.java:40]: - This is not really optimal to copy byte by byte.
- IOStreams should be closed (using try-with-resources block)

To sum up: you can use https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/IOUtils.html#copy-java.io.InputStream-java.io.OutputStream-

Here and in two more places

File: core/src/main/java/greencity/entity/SocialNetwork.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+public class SocialNetwork {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Length(min = 1, max = 500)
+    @Column(name = "social_network_url", length = 500)
+    String url;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (antonbakalets) [inline core/src/main/java/greencity/entity/SocialNetwork.java:23]: I don't think it should be LAZY from this side of the connection.

File: core/src/main/java/greencity/entity/SocialNetwork.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    User user;
+
+    @OneToOne(fetch = FetchType.LAZY)
+    SocialNetworkImage socialNetworkImage;
```
Comment (antonbakalets) [inline core/src/main/java/greencity/entity/SocialNetwork.java:27]: This one also not really LAZY as almost always you will be loading user social networks with corresponding image.

File: core/src/main/java/greencity/service/impl/SocialNetworkServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        List<SocialNetwork> listSocialNetwork = socialNetworkUrls.stream()
+            .map(url ->
+                SocialNetwork.builder()
+                    .url(url)
+                    .user(user)
+                    .socialNetworkImage(socialNetworkImageService.getSocialNetworkImageByUrl(url))
+                    .build()
+            )
+            .collect(Collectors.toList());
+        socialNetworkRepo.deleteAllByUserIs(user);
```
Comment (antonbakalets) [inline core/src/main/java/greencity/service/impl/SocialNetworkServiceImpl.java:39]: Can you please check if the following will work instead (please check logged SQL statements):
```
user.getSocialNetworks().clear();
user.setSocialNetworks(listSocialNetwork);
```
If that is the case you will be able to remove those Repos. Sometime JPA can do some magic...

**Review by antonbakalets — CHANGES_REQUESTED** [review]

File: core/src/main/java/greencity/entity/SocialNetwork.java (hunk 3 above)
Comment (vorfolomey) [reply core/src/main/java/greencity/entity/SocialNetwork.java:23]: What does that mean?

File: core/src/main/java/greencity/entity/SocialNetwork.java (hunk 4 above)
Comment (vorfolomey) [reply core/src/main/java/greencity/entity/SocialNetwork.java:27]: Okay

**Review by Yura09 — APPROVED** [review]

File: core/src/main/java/greencity/service/impl/SocialNetworkServiceImpl.java (hunk 5 above)
Comment (vorfolomey) [reply core/src/main/java/greencity/service/impl/SocialNetworkServiceImpl.java:39]: Попробував. Не працює чогось. Тільки в об'єкті зміни, але в бд нічого не міняється

**Review by TarasDovganyuk — APPROVED** [review]

**Review by MarDatsko — APPROVED** [review]

**Review by eyouu — APPROVED** [review]

File: core/src/main/java/greencity/service/impl/SocialNetworkServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        List<SocialNetwork> listSocialNetwork = socialNetworkUrls.stream()
+            .map(url ->
+                SocialNetwork.builder()
+                    .url(url)
+                    .user(user)
+                    .socialNetworkImage(socialNetworkImageService.getSocialNetworkImageByUrl(url))
+                    .build()
+            )
+            .collect(Collectors.toList());
+        user.getSocialNetworks().clear();
```
Comment (antonbakalets) [inline core/src/main/java/greencity/service/impl/SocialNetworkServiceImpl.java:37]: Please move those two lines out of this method (one level up), then:

- remove parameter user

**Review by antonbakalets — APPROVED** [review]

**Pull request merged**